Anti Xray: Obfuscate blocks touching lava feature

stonar96 opened the pull request:
Resolves #2940.
Obfuscating blocks touching lava can be disabled by setting the new config
option anti-xray.lava-obscures in the paper world config to false.
Clearing lava with a bucket or replacing it by a transparent block already
triggers block updates for nearby blocks.

stonar96 committed:
Obfuscate blocks touching lava

Resolves #2940.
Obfuscating blocks touching lava can be disabled by setting the new config
option anti-xray.lava-obscures in the paper world config to false.
Clearing lava with a bucket or replacing it by a transparent block already
triggers block updates for nearby blocks.

Proximyst commented on Spigot-Server-Patches/0384-Anti-Xray.patch:
```diff
@@ -28,6 +28,7 @@ index 48676152152faf7a7b9524ac37d8b4a8c32c4e2c..ca2ac17747d4f2acf1df056759c5c182
 +    public EngineMode engineMode;
```
Use default values here and reference the variable as defaults when getting the values.

EDIT: You apparently can't resolve conversations in reviews, so ignore this :)

zachbr replied:
There is no distinction, it is author preference. Given the rest of antixray config is written this way I see no reason he needs to change it now.

Proximyst replied:
Was under the impression that was the "standard" :p

stonar96 replied:
Not sure why this is necessary suddenly. Most other config options also don't do this. Also this can be done in another PR if needed. Edit: didn't reload recent comments.

aikar replied:
as said its not a standard, either way is fine. I personally do mine with the other pattern as sometimes what default you use in the method code might vary (for migrations).

Additionally, the method might be a little more complex and big, so I find it easier to view the defaults in the initializer of the variable.

but no reason to update pr over this, both is valid.

Proximyst commented on Spigot-Server-Patches/0384-Anti-Xray.patch:
```diff
@@ -38,6 +39,7 @@ index 48676152152faf7a7b9524ac37d8b4a8c32c4e2c..ca2ac17747d4f2acf1df056759c5c182
 +        maxChunkSectionIndex = getInt("anti-xray.max-chunk-section-index", 3);
 +        maxChunkSectionIndex = maxChunkSectionIndex > 15 ? 15 : maxChunkSectionIndex;
 +        updateRadius = getInt("anti-xray.update-radius", 2);
++        lavaObscures = getBoolean("anti-xray.lava-obscures", true);
```
Is changing default behaviour after so many versions a good idea?

stonar96 replied:
Aikar on Discord: lava-obscures: true, default on is reasonable here

aikar replied:
Yes, defaults represent what majority will want. majority will clearly want this.

stonar96 committed:
Set lava-obscures to false by default

Server owners should also modify their hidden-blocks list when enabling this
option, such that it only contains ores with stone texture at the upper edges.

VADemon commented on Spigot-Server-Patches/0384-Anti-Xray.patch:
```diff
@@ -208,7 +210,8 @@ index 0000000000000000000000000000000000000000..9c017eb027ce52ea6a2414af2d0190b3
 +                solidGlobal[i] = blockData.getBlock().isOccluding(blockData, emptyChunk, zeroPos)
-+                    && blockData.getBlock() != Blocks.SPAWNER && blockData.getBlock() != Blocks.BARRIER && blockData.getBlock() != Blocks.SHULKER_BOX;
++                    && blockData.getBlock() != Blocks.SPAWNER && blockData.getBlock() != Blocks.BARRIER && blockData.getBlock() != Blocks.SHULKER_BOX || paperWorldConfig.lavaObscures && blockData == Blocks.LAVA.getBlockData();
++                // Comparing blockData == Blocks.LAVA.getBlockData() instead of blockData.getBlock() == Blocks.LAVA ensures that only "stationary lava" is used
```
Comparing against stationary lava is not enough. The top-most stationary lava block will reveal 2 top texels (not fill the block fully).

There are 2 ores that are visible this way using the default vanilla texture pack: Emerald and Nether Quartz ore. Not to mention custom modpacks that may make other ores visible in this case by drawing ore textures closer to the block edges.

Whether or not custom legitimate texture packs should be considered is up to debate. But I believe current patch will wrongly hide Emerald/Quartz ore, if so - it's hiding otherwise visible ores.

Proximyst replied:
The top 2 pixels was brought up on Discord, and ultimately it was disregarded as changing how anti x-ray works at all is not worth it for a not so common issue. You're always able to disable this feature and rather just use moderators.

stonar96 replied:
This is known and has already been discussed. (See #2940.) Won't (and can't) fix. Server owners must decide if they want this feature and also take look at the notes in the commit messages.
